### PR TITLE
mlton: refactor and adds mlton20180207Binary mlton20180207 mltonHEAD

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Newer versions of the MLton compiler in nixpkgs.

###### Things done

Adds the following derivations:

* mlton20130715
  - for this one I just copy and pasted the old default.nix
  - I have tested some projects that compile with this version and don't
    compile with the newer version, so I think it makes sense to keep
    the old version

* mlton20180207Binary
  - This is used to build the source derivations

* mlton20180207
  - latest release

* mltonHEAD
  - latest commit to the master branch of the github repo at the time of
    creating this derivation

Everything seemed to build fine using Nix 2.0 on NixOS. I haven't tried building it on macOS.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/mlton/20130715.nix [hunk 1]
```diff
@@ -0,0 +1,119 @@
+{ stdenv, fetchurl, patchelf, gmp }:
+
+let
+  version = "20130715";
+
+  usr_prefix = if stdenv.isDarwin then "usr/local" else "usr";
+
+  dynamic_linker =
```
[the remark this answers is not in the document]
Comment (srdqty) [reply pkgs/development/compilers/mlton/20130715.nix]: Nice, I'll change that.

**Comment by joachifm** [comment]
@GrahamcOfBorg build mlton mltonHEAD

**Comment by srdqty** [comment]
Ok, I see `/usr/bin/env: bad interpreter: No such file or directory` errors in the x86_64-linux build for files that have `/usr/bin/env bash` shebang lines. I should be able to fix that pretty easily. I'll test with sandboxing this time.

Looks like the x86_64-darwin build succeeded, but ran out of time when it was running tests. Not super surprising, building mlton is pretty intensive.

**Comment by jasoncarr0** [comment]
Are you still working on fixing this? The /usr/bin/env bash lines should be able to be fixed by the same
`find . -type f | grep -v -e '\.tgz''$' | xargs sed -i "s@/usr/bin/env bash@$(type -p bash)@"` line that was in the older version. For the timeout, I'm not aware of whether there's a way to set that in mlton. Do you think it would be a good idea to skip tests?

**Comment by srdqty** [comment]
Yeah I'm going to update this.

**Comment by matthewbauer** [comment]
Thanks!

**Comment by jasoncarr0** [comment]
This was merged, but @srdqty can you actually list which projects don't compile with the new version but compiled with the old?

**Comment by srdqty** [comment]
Sure. It was a personal project: https://github.com/srdqty/sexpr-pp

Seems that you cannot redefine the `=` operator in the new version. I'm not sure which case is correct according to the SML definition. This project wasn't really idiomatic code, so maybe it doesn't affect many projects.

```
λx. make
mlton src/sexpr-repl.mlb
Error: src/symbol-table/symbol.sig 8.5-8.5.
  Special identifier cannot be specified by val: =.
    in: val = : ((t * t) -> bool)
Error: src/symbol-table/symbol-table.fun 39.10-39.10.
  Special identifier cannot be redefined by fun: =.
    in: fun op = (a, b) = (eq (a, b))
Error: src/symbol-table/symbol.sig 8.5-8.5.
  Special identifier cannot be specified by val: =.
    in: val = : ((t * t) -> bool)
Error: src/lexical-context/mark/mark.sig 7.7-7.7.
  Special identifier cannot be specified by val: =.
    in: val = : ((t * t) -> bool)
Error: src/lexical-context/mark/mark.sig 7.7-7.7.
  Special identifier cannot be specified by val: =.
    in: val = : ((t * t) -> bool)
Error: src/lexical-context/mark/mark.fun 17.10-17.10.
  Special identifier cannot be redefined by fun: =.
    in: fun op = (m1, m2) = (eq (m1, m2))
Error: src/symbol-table/symbol.sig 8.5-8.5.
  Special identifier cannot be specified by val: =.
    in: val = : ((t * t) -> bool)
unhandled exception: Fail: compilation aborted: parseAndElaborate reported errors
make: *** [Makefile:5: sexpr-repl] Error 1
```

**Comment by jasoncarr0** [comment]
So, it turns out that while that's supported by SMLNJ, it is expressly forbidden by the spec. http://mlton.org/SMLNJDeviations

What I'd like to eventually do is be able to get rid of 2013, so I was wondering about any deviations between the versions.

**Pull request merged**